src/ctl/src/cmd_impl/hummock/validate_version.rs

```diff
@@ -30,3 +46,177 @@ pub async fn validate_version(context: &CtlContext) -> anyhow::Result<()> {
 
     Ok(())
 }
+
+async fn get_archive(
+    archive_id: HummockVersionId,
+    data_dir: &str,
+    archive_object_store: ObjectStoreRef,
+) -> anyhow::Result<HummockVersionArchive> {
+    use prost::Message;
+    let archive_dir = version_archive_dir(data_dir);
+    let archive_path = format!("{archive_dir}/{archive_id}");
+    let archive_bytes = archive_object_store.read(&archive_path, ..).await?;
+    let archive: HummockVersionArchive = HummockVersionArchive::decode(archive_bytes)?;
+    Ok(archive)
+}
+
+pub async fn print_user_key_in_archive(
+    context: &CtlContext,
+    archive_ids: Vec<HummockVersionId>,
+    data_dir: String,
+    user_key: String,
+) -> anyhow::Result<()> {
+    let user_key_bytes = hex::decode(user_key.clone()).unwrap_or_else(|_| {
+        panic!("cannot decode user key {} into raw bytes", user_key);
+    });
+    let user_key = UserKey::decode(&user_key_bytes);
+    println!("user key: {user_key:?}");
+
+    let hummock_opts = HummockServiceOpts::from_env(Some(data_dir.clone()))?;
+    let hummock = context.hummock_store(hummock_opts).await?;
+    let sstable_store = hummock.sstable_store();
+    let archive_object_store = sstable_store.store();
+    for archive_id in archive_ids.into_iter().sorted() {
+        println!("search archive {archive_id}");
+        let archive = get_archive(archive_id, &data_dir, archive_object_store.clone()).await?;
+        let mut base_version =
+            HummockVersion::from_persisted_protobuf(archive.version.as_ref().unwrap());
+        print_user_key_in_version(sstable_store.clone(), &base_version, &user_key).await?;
+        for delta in &archive.version_deltas {
+            base_version.apply_version_delta(&HummockVersionDelta::from_persisted_protobuf(delta));
+            print_user_key_in_version(sstable_store.clone(), &base_version, &user_key).await?;
+        }
+    }
+    Ok(())
+}
+
+async fn print_user_key_in_version(
+    sstable_store: SstableStoreRef,
+    version: &HummockVersion,
+    target_key: &UserKey<&[u8]>,
+) -> anyhow::Result<()> {
+    println!("print key {:?} in version {}", target_key, version.id);
+    for cg in version.levels.values() {
+        for level in cg
+            .l0
+            .as_ref()
+            .unwrap()
+            .sub_levels
+            .iter()
+            .rev()
+            .chain(cg.levels.iter())
+        {
+            for sstable_info in &level.table_infos {
+                use risingwave_hummock_sdk::key_range::KeyRange;
+                let key_range: KeyRange = sstable_info.key_range.as_ref().unwrap().into();
+                let left_user_key = FullKey::decode(&key_range.left);
+                let right_user_key = FullKey::decode(&key_range.right);
+                if left_user_key.user_key > *target_key || *target_key > right_user_key.user_key {
```

Comment:
should we need to use `KeyComparator`?

Reply:
It seem no method that compares user key with key range, or user key with user key?